evm-precompiles/erc20/src/lib.rs

```diff
@@ -378,29 +357,65 @@ where
 		let to: H160 = input.read::<Address>(gasometer)?.into();
 		let amount: Balance = input.read::<U256>(gasometer)?.saturated_into();
 
-		{
-			let caller: Runtime::AccountId = Runtime::AddressMapping::into_account_id(context.caller);
-			let from: Runtime::AccountId = Runtime::AddressMapping::into_account_id(from.clone());
-			let to: Runtime::AccountId = Runtime::AddressMapping::into_account_id(to);
+		// If caller is "from", it can spend as much as it wants from its own balance.
+		if context.caller != from {
+			// Dispatch call (if enough gas).
+			gasometer.record_cost(RuntimeHelper::<Runtime>::db_read_gas_cost())?;
+
+			let current_approved_amount: Balance =
+				crml_token_approvals::Module::<Runtime>::erc20_approvals((&from.clone(), &asset_id), &context.caller);
+			if current_approved_amount >= amount {
```

Comment:
would do sth like this to reduce the nesting
```suggestion
			let new_approved_amount: Balance = current_approved_amount.checked_sub(amount).ok_or(error("caller not approved").into())?;
```